### PR TITLE
Fixed auto-lock while recording voice message.

### PR DESCRIPTION
**Pull request by 23rd**
This PR fixes [this one](https://github.com/telegramdesktop/tdesktop/issues/5695).
If you don't mind, I also added some `const` to vars within `checkAutoLock()`.
If this is not good for single commit, I'll remove it.

### CODE REVIEW

**Comment by 23rd** [comment]
I have realized that it's better to use `std::min()` instead of `qMin()`.

**Comment by john-preston** [comment]
Thanks for the fix. I think this can be improved in terms of code meaning. All the autolock delays are due to the computer not being idle. So all those are `idleFor` really.

I suggest changing `psIdleTime` to `Platform::LastUserInputTime` and making one `crl::time` value `_lastNonIdleTime` in `Application` (instead of `lastTime*At` variables in AuthSessionSettings) + add `[[nodiscard]] crl::time lastNonIdleTime() const` and `void updateNonIdle();` in `Application`, the first will return `std::max(Platform::LastUserInputTime(), _lastNonIdleTime)`, the second will set `_lastNonIdleTime` to `crl::now()`.

Because we really are not interested when was the last update in video playback or voice recording. And also they really are not current user settings in any sense, but the last non idle time is something about the whole application (and the platform-specific last user input time is taken even from the whole system, not single application, where there is platform API to do that).

That way both `Auth().settings().setLastTime*At(crl::now())` will become just `Core::App().updateNonIdle()`.

**Comment by john-preston** [comment]
Maybe the same logic can be applied to other places where `psIdleTime` is used right now. For example where the notifications are delayed on the screen until the user input happens - it will be good if they won't be delayed in case video is watched or voice message is recorded, so `Core::App().lastNonIdleTime()` can be used there too.

**Comment by 23rd** [comment]
@john-preston 

Thanks for replies.
Unfortunately, the `std::max(Platform::LastUserInputTime(), _lastNonIdleTime)` does not work in the way that you suggested. I replaced it with `return _lastNonIdleTime;` and it came to work as it should.

>For example where the notifications are delayed on the screen until the user input happens - it will be good if they won't be delayed in case video is watched or voice message is recorded, so `Core::App().lastNonIdleTime()` can be used there too.

Yeah, I was surprised to find that checking of the last user input properly works only for Windows. And it repeats the code of the `Platform::LastUserInputTime()` from `specific_win.h`, so I replaced it with a hope that it will work for all platforms. I splited it into 2 commits, because notification delay is a little bit another story.

Count of changed files is huge now. =\

**Comment by john-preston** [comment]
Why it doesn't work? I guess, 'LastUserInputTime' in platform-specific code is not on the same scale / origin, as crl::now()? But it could be fixed by LastUserInputTime returning `crl::now() - psIdleTime()` for the old implementation of `psIdleTime`.

**Comment by john-preston** [comment]
Last time I checked it worked properly on macOS as well. Maybe they've changed something in the API since then, I'm not sure.

**Comment by john-preston** [comment]
@23rd UPD I see now. As I understand you've renamed `psIdleTime` to `LastUserInputTime`, but that was not what I've suggested. `LastUserInputTime` was supposed to return the last time there was an input event, so in terms of old code `LastUserInputTime() { return crl::now() - psIdleTime(); }`, but removing `psIdleTime` and leaving the related code just in `LastUserInputTime` and using it in `Application::lastNonIdleTime` - max of the last time user input was detected by platform-specific APIs, last time a video was played or last time a voice message was recorded.

File: Telegram/SourceFiles/platform/win/specific_win.cpp [hunk 1]
```diff
@@ -350,6 +344,12 @@ QString CurrentExecutablePath(int argc, char *argv[]) {
 	return QString();
 }
 
+crl::time LastUserInputTime() {
+	LASTINPUTINFO lii;
+	lii.cbSize = sizeof(LASTINPUTINFO);
+	return GetLastInputInfo(&lii) ? (GetTickCount() - lii.dwTime) : (crl::now() - _lastUserAction);
```
Comment (john-preston) [inline Telegram/SourceFiles/platform/win/specific_win.cpp]: So, this one will be `GetLastInputInfo(&lii) ? (crl::now() + lii.dwTime - GetTickCount()) : _lastUserAction`.

File: Telegram/SourceFiles/mainwidget.cpp [hunk 2]
```diff
@@ -3767,7 +3767,7 @@ void MainWidget::updateOnline(bool gotOtherOffline) {
 	bool isOnline = !App::quitting() && App::wnd()->isActive();
 	int updateIn = Global::OnlineUpdatePeriod();
 	if (isOnline) {
-		auto idle = psIdleTime();
+		const auto idle = Platform::LastUserInputTime();
```
Comment (john-preston) [inline Telegram/SourceFiles/mainwidget.cpp]: Here it'll be `idle = crl::now() - Platform::LastUserInputTime()`.

File: Telegram/SourceFiles/mainwidget.cpp [hunk 3]
```diff
@@ -3883,7 +3883,7 @@ void MainWidget::writeDrafts(History *history) {
 }
 
 void MainWidget::checkIdleFinish() {
-	if (psIdleTime() < Global::OfflineIdleTimeout()) {
+	if (Platform::LastUserInputTime() < Global::OfflineIdleTimeout()) {
```
Comment (john-preston) [inline Telegram/SourceFiles/mainwidget.cpp]: Here it'll be `crl::now() - Platform::LastUserInputTime() < Global::OfflineIdleTimeout()`.

File: Telegram/SourceFiles/platform/linux/specific_linux.cpp [hunk 4]
```diff
@@ -449,6 +445,10 @@ bool OpenSystemSettings(SystemSettingsType type) {
 	return true;
 }
 
+crl::time LastUserInputTime() {
+	return crl::now() - _lastUserAction;
```
Comment (john-preston) [inline Telegram/SourceFiles/platform/linux/specific_linux.cpp]: Here it'll be just `_lastUserAction`.

File: Telegram/SourceFiles/platform/mac/specific_mac.mm [hunk 5]
```diff
@@ -104,11 +104,6 @@ bool psIdleSupported() {
 	return objc_idleSupported();
 }
 
-crl::time psIdleTime() {
-	auto idleTime = 0LL;
-	return objc_idleTime(idleTime) ? idleTime : (crl::now() - _lastUserAction);
```
Comment (john-preston) [inline Telegram/SourceFiles/platform/mac/specific_mac.mm]: Here it'll be `objc_idleTime(idleTime) ? (crl::now() - crl::time(idleTime)) : _lastUserAction`.

File: Telegram/SourceFiles/platform/mac/specific_mac.mm [hunk 6]
```diff
@@ -270,6 +265,11 @@ bool OpenSystemSettings(SystemSettingsType type) {
 	return true;
 }
 
+crl::time LastUserInputTime() {
+	auto idleTime = 0LL;
+	return objc_idleTime(idleTime) ? idleTime : (crl::now() - _lastUserAction);
```
Comment (john-preston) [inline Telegram/SourceFiles/platform/mac/specific_mac.mm]: Here it'll be `objc_idleTime(idleTime) ? (crl::now() - crl::time(idleTime)) : _lastUserAction`.

File: Telegram/SourceFiles/window/notifications_manager_default.cpp [hunk 7]
```diff
@@ -588,14 +586,8 @@ void Notification::prepareActionsCache() {
...
 
-	auto wasUserInput = true; // TODO
-#ifdef Q_OS_WIN
-	LASTINPUTINFO lii;
-	lii.cbSize = sizeof(LASTINPUTINFO);
-	BOOL res = GetLastInputInfo(&lii);
-	wasUserInput = (!res || lii.dwTime >= _started);
-#endif // Q_OS_WIN
-	if (wasUserInput) {
+	if ((crl::now() - Platform::LastUserInputTime() > _started)
```
Comment (john-preston) [inline Telegram/SourceFiles/window/notifications_manager_default.cpp]: Here it'll be just `Core::App().lastNonIdleTime() > _started`.

File: Telegram/SourceFiles/auth_session.cpp [hunk 8]
```diff
@@ -483,11 +483,9 @@ void AuthSession::checkAutoLock() {
...
 
 	Core::App().checkLocalTime();
-	auto now = crl::now();
-	auto shouldLockInMs = Global::AutoLock() * 1000LL;
-	auto idleForMs = psIdleTime();
-	auto notPlayingVideoForMs = now - settings().lastTimeVideoPlayedAt();
-	auto checkTimeMs = qMin(idleForMs, notPlayingVideoForMs);
+	const auto now = crl::now();
+	const auto shouldLockInMs = Global::AutoLock() * 1000LL;
+	const auto checkTimeMs = std::min(Platform::LastUserInputTime(), now - Core::App().lastNonIdleTime());
```
Comment (john-preston) [inline Telegram/SourceFiles/auth_session.cpp]: Here it'll be just `now - Core::App().lastNonIdleTime()`.

**Comment by john-preston** [comment]
@23rd UPD I see now about the `only on Windows` - that's because on Linux I didn't find such API and on macOS (at least starting with OS X 10.8) only native notifications are used, so the code of custom notifications that are kept on the screen while there is no user input is never used anyway. But it is a good idea to use the same methods as in other places, that's good.

**Comment by 23rd** [comment]
Oh...
To be honest, I don't see much advantage, we do subtraction in one place, but add in another.

**Comment by 23rd** [comment]
@john-preston 
The code updated.

Sorry for misunderstanding about `psIdleTime` and `LastUserInputTime`.

**Comment by john-preston** [comment]
@23rd 

> we do subtraction of crl::now() in one place, but add it in another.

Yes, this is not really a simplification in the amount of operations / code, but a more consistent naming - one platform-specific time value - from which point of time we think there was no user attention to the app (last user input time) + one in-app value - with the same idea (last time we've updated voice message recording or video frame was shown). And one application-wide method, that returns max of those two values.

File: Telegram/SourceFiles/window/notifications_manager_default.cpp [hunk 9]
```diff
@@ -588,14 +586,7 @@ void Notification::prepareActionsCache() {
...
 
-	auto wasUserInput = true; // TODO
-#ifdef Q_OS_WIN
-	LASTINPUTINFO lii;
-	lii.cbSize = sizeof(LASTINPUTINFO);
-	BOOL res = GetLastInputInfo(&lii);
-	wasUserInput = (!res || lii.dwTime >= _started);
-#endif // Q_OS_WIN
-	if (wasUserInput) {
+	if ((Core::App().lastNonIdleTime() > _started)) {
```
Comment (john-preston) [inline Telegram/SourceFiles/window/notifications_manager_default.cpp]: No need for double parentheses here? Just `if (Core::App().lastNonIdleTime() > _started) {`?

**Comment by john-preston** [comment]
No, not the last one. Now I see one more way to simplify this, sorry.

Right now all three platform-specific code paths track `_lastUserAction` time in `psUserActionDone` call. Really this call can be made no-op everywhere except Windows version (where it does something else too) and we should just call `Core::App().updateNonIdle()` where we now call `psUserActionDone()`.

That way `_lastUserAction` is not needed at all, it will be tracked inside `_lastNonIdleTime` together with video watching and voice message recording. And on Linux `Platform::LastUserInputTime()` will simply return zero, because there is no API for that (other systems will return value if they could get it and zero otherwise).

**Comment by john-preston** [comment]
I can be `Core::App().userActionDone()` that'll call both `psUserActionDone()` and `updateNonIdle()`.

It is possible that `psUserActionDone` is not needed at all, even on Windows, but right now I'm not sure, so I'd leave it as it is.

**Comment by 23rd** [comment]
In that case, probably should not hurry to merge this PR, I would test it a little more.

https://github.com/telegramdesktop/tdesktop/blob/e174025a92c67f5a5c6545fbcda6e44dbd7f3348/Telegram/SourceFiles/platform/win/specific_win.cpp#L137-L140

As I can see, this call is just to be sure that the sound of notifications would not be skipped after the log off or the lock.

**Comment by john-preston** [comment]
Yes, it should work even without that, but to be sure that we are in the right state in case of any user input we force us to be not in session-logged-out mode.

**Comment by 23rd** [comment]
@john-preston 

So, I added few `LOG()` in `case WM_WTSSESSION_CHANGE: {` and tested lock in Windows 7, Windows 8.1 and Windows 10. Everywhere handler works well, it is called even twice.

If we still want to be sure that we a not logged off, I suggest to replace
```
if (wParam == WTS_SESSION_LOGOFF || wParam == WTS_SESSION_LOCK) {
	setSessionLoggedOff(true);
} else if (wParam == WTS_SESSION_LOGON || wParam == WTS_SESSION_UNLOCK) {
	setSessionLoggedOff(false);
}
```
with
```
setSessionLoggedOff(wParam == WTS_SESSION_LOGOFF || wParam == WTS_SESSION_LOCK);
```

But I see no reason not to trust windows event handler.

**Comment by john-preston** [comment]
@23rd Yes, perhaps we can check if that works fine.

**Comment by john-preston** [comment]
@23rd I mean, try without handling user input for that and see if there'll be complaints.

**Comment by 23rd** [comment]
I don't get you.
In the latest commit I removed `psUserActionDone()` with `EventFilter::getInstance()->setSessionLoggedOff(false);` and a little tested it.
In Windows with locked user the sound of notifications was skipped, but when I loged in the sound returned.

**Comment by john-preston** [comment]
@23rd Yes, I've meant exactly that - try that version (that you already pushed, I didn't see that, was replying to your comment) in production and see if there are any complaints. I think there won't be any and everything is fine. Thanks!

**Pull request merged**